### PR TITLE
enhancement(Cloud Databases): add support for User types, roles 

### PR DESCRIPTION
**Pull request by alexhemard**
Adds support for multiple user types (`database`, `ops_manager`, `read_only_replica`) for the IBM Database Resource 
See https://cloud.ibm.com/apidocs/cloud-databases-api/cloud-databases-api-v5#createdatabaseuser

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Output from acceptance testing: 

```sh
$ make testacc TESTARGS='-run=TestAccIBMMongoDBEnterprise'
...
--- PASS: TestAccIBMMongoDBEnterpriseDatabaseInstanceGroupBasic (5207.81s)
--- PASS: TestAccIBMMongoDBEnterpriseDatabaseInstanceBasic (27551.47s)
...
$ make testacc TESTARGS='-run=TestAccIBMDatabaseInstance'
...
--- PASS: TestAccIBMDatabaseInstanceElasticsearchImport (465.51s)
--- PASS: TestAccIBMDatabaseInstanceEtcdImport (870.41s)
--- PASS: TestAccIBMDatabaseInstanceMongodbBasic (1534.00s)
--- PASS: TestAccIBMDatabaseInstanceMongodbImport (1152.09s)
--- PASS: TestAccIBMDatabaseInstancePostgresBasic (2373.14s)
--- PASS: TestAccIBMDatabaseInstancePostgresGroup (2593.68s)
--- PASS: TestAccIBMDatabaseInstancePostgresImport (944.31s)
--- PASS: TestAccIBMDatabaseInstancePostgresNode (2270.03s)
--- PASS: TestAccIBMDatabaseInstanceRabbitmqImport (809.02s)
--- PASS: TestAccIBMDatabaseInstanceRedisImport (661.52s)
--- PASS: TestAccIBMDatabaseInstance_Cassandra_Group (6637.37s)
--- PASS: TestAccIBMDatabaseInstance_Cassandra_Node (4408.87s)
--- PASS: TestAccIBMDatabaseInstance_Elasticsearch_Basic (854.69s)
--- PASS: TestAccIBMDatabaseInstance_Elasticsearch_Group (1522.81s)
--- PASS: TestAccIBMDatabaseInstance_Elasticsearch_Node (789.75s)
--- PASS: TestAccIBMDatabaseInstance_Etcd_Basic (873.66s)
--- PASS: TestAccIBMDatabaseInstance_Rabbitmq_Basic (1422.83s)
--- PASS: TestAccIBMDatabaseInstance_Redis_Basic (705.72s)
...
```

### CODE REVIEW

File: ibm/config.go [hunk 1]
```diff
@@ -1732,6 +1743,18 @@ func (c *Config) ClientSession() (interface{}, error) {
 	}
 	session.icdServiceAPI = icdAPI
```
Comment (hkantare) [inline ibm/config.go:1745]: Add support to construct the private endpoint if the visibility is set to private

File: ibm/data_source_ibm_database.go [hunk 2]
```diff
@@ -117,6 +117,12 @@ func dataSourceIBMDatabaseInstance() *schema.Resource {
 							Computed:    true,
 							Sensitive:   true,
 						},
+						"user_type": {
```
Comment (hkantare) [inline ibm/data_source_ibm_database.go:120]: In datasource it can't be required it should be computed

File: ibm/resource_ibm_database.go [hunk 3]
```diff
@@ -1251,23 +1262,40 @@ func resourceIBMDatabaseInstanceCreate(d *schema.ResourceData, meta interface{})
...
+
+			instanceId := d.Id()
+			createDatabaseUserOptions := &clouddatabasesv5.CreateDatabaseUserOptions{
+				ID:       &instanceId,
+				UserType: core.StringPtr(userEl["type"].(string)),
+				User:     createDatabaseUserRequestUserModel,
 			}
-			task, err := icdClient.Users().CreateUser(icdId, userReq)
+
+			fmt.Printf("User: %v", createDatabaseUserOptions)
```
Comment (hkantare) [inline ibm/resource_ibm_database.go:1286]: Remove this fmt.Printf

File: ibm/resource_ibm_database.go [hunk 4]
```diff
@@ -1721,63 +1755,77 @@ func resourceIBMDatabaseInstanceUpdate(d *schema.ResourceData, meta interface{})
...
+				createDatabaseUserOptions := &clouddatabasesv5.CreateDatabaseUserOptions{
+					ID:       &instanceID,
+					UserType: core.StringPtr(newEntry["type"].(string)),
+					User:     userEntry,
 				}
-				userReq := icdv4.UserReq{
-					User: userEntry,
+
+				createDatabaseUserResponse, _, err := cloudDatabasesV5.CreateDatabaseUser(createDatabaseUserOptions)
+				if err != nil {
```
Comment (hkantare) [inline ibm/resource_ibm_database.go:1795]: We are returning the error if we have any error due to createDatabse the line 1802 will never be executed if the user already exists and update the password

File: website/docs/r/database.html.markdown [hunk 5]
```diff
@@ -361,17 +366,19 @@ Review the argument reference that you can specify for your resource.
 - `users` - (Optional, List of Objects) A list of users that you want to create on the database. Multiple blocks are allowed.
 
   Nested scheme for `users`:
-  - `name` - (Optional, String) The user ID to add to the database instance. The user ID must be in the range 5 - 32 characters.
-  - `password` - (Optional, String) The password for the user ID. The password must be in the range 10 - 32 characters.
+  - `name` - (Optional, String) The user name to add to the database instance. The user name must be in the range 5 - 32 characters.
+  - `password` - (Optional, String) The password for the user. The password must be in the range 10 - 32 characters.
+  - `type` - (Required, String) The type for the user. Examples: database, ops_manager, read_only_replica
```
Comment (hkantare) [inline website/docs/r/database.html.markdown:371]: type is optional with default value ...

**Comment by hkantare** [comment]
The doc for database datasource is not updated website/docs/d/database.html.markdown 

Please share the testresults of all the testcases...

**Comment by alexhemard** [comment]
> The doc for database datasource is not updated website/docs/d/database.html.markdown

Removed the changes to the  `ibm_database` data source but updated the documentation since it currently seems to imply it will create a Read-only replica

**Comment by hkantare** [comment]
Can you reabse the PR with respective to new folder structure

File: ibm/service/database/resource_ibm_database.go [hunk 6]
```diff
@@ -1488,23 +1525,38 @@ func resourceIBMDatabaseInstanceCreate(context context.Context, d *schema.Resour
...
+			createDatabaseUserRequestUserModel := &clouddatabasesv5.User{
+				Username: core.StringPtr(userEl["name"].(string)),
+				Password: core.StringPtr(userEl["password"].(string)),
 			}
-			task, err := icdClient.Users().CreateUser(icdId, userReq)
+
+			instanceId := d.Id()
+			createDatabaseUserOptions := &clouddatabasesv5.CreateDatabaseUserOptions{
+				ID:       &instanceId,
+				UserType: core.StringPtr(userEl["type"].(string)),
```
Comment (hkantare) [inline ibm/service/database/resource_ibm_database.go:1544]: Why we are not we considering role during creation process?

Comment (alexhemard) [reply ibm/service/database/resource_ibm_database.go:1549]: I missed this :worried: Good catch!

**Pull request merged**